design-history/exploration-reports/2019.06-unixfsv2-spike-01.md

```diff
@@ -0,0 +1,296 @@
+A unixfsv2 spike, using schemas
+===============================
+
+This document is an exploration of what a hypothetical "unixfs v2" might look like,
+using schemas as a design language for describing the data structures.
+
+This is not a final proposal, but rather a lens to explore both unixfsv2
+concepts and schemas themselves at the same time, and how we expect them
+to contribute to each other.  Several concepts in schemas, advanced layouts,
+their interactions, and their syntaxes will be trialled here; they are not final.
+
+Without further ado...
+
+Schema
+------
+
+```
+type AnyFile union {
+	| File "f"
+	| Dir "d"
+	| Symlink "s"
+} representation keyed
+
+type Filename string
+
+type Dir struct {
+	members {Filename:DirEnt}<HAMT>
+}
+
+type DirEnt struct {
+	content &AnyFile
+	attribs Attribs
+}
+
+type File struct {
+	content Bytes<Rabin>
+}
+
+type Symlink struct {
+	target String
+}
+
+advanced Rabin {
+	kind bytes
+	implementation "IPLD/experimental/rabin/v1"
+}
+
+advanced HAMT {
+	kind map
+	implementation "IPLD/experimental/HAMT/v1"
+}
+
+type Attribs struct {
+	mtime Int # Danger danger will robinson; int sizes will rear their ugly head.
+	posix Int # Can we just... do the standard 0777 mask packing here?  I don't think anything else is *less* work.
+	sticky Bool (implicit: false)
+	setuid Bool (implicit: false)
+	setgid Bool (implicit: false)
+	uid Int # make another schema which deliberately excludes this.  try to match this one (the one with more info) first.
+	gid Int # make another schema which deliberately excludes this.  try to match this one (the one with more info) first.
+}
+
+# You should be able to imagine sending a fancy graphsync request with some structure like:
+#
+#  {"schema": {{schemaCIDofUnixfs}}, "selector": {{selectorSpec}},
+#   "schemaPatch": {"./dir/transparently/through/to/file": {{schemaCIDofRabinLayout}},
+#   "startAt": {{rootBlockCID+path}} }
+#
+# This kind of fancy graphsync request can be used to make a query which does "the right thing" with
```

Comment:
I started this “File and Dir *as* composites” experiment and it’s quite nice. It literally **forces** us to store the configuration that generated the file which should make it a lot easier to ensure different implementations produce consistent hashes.

* [File](https://github.com/ipld/js-unixfsv2/blob/types/src/file.js)
* [Dir](https://github.com/ipld/js-unixfsv2/blob/types/src/dir.js)

One really cool thing we get here is that the metadata is easily exposed on File objects as `/Filename.txt/guid`. However, I can’t think of the ideal way to:

* Expose the same metadata on directories since the namespace will conflict with files in the directory.
* Expose a property that returns **all** metadata.

Any property you pick will end up being reserved so it can’t be used for arbitrary metadata or as a filename. The one property we *could* use safely since it’s already reserved in filesystems is “.” but then our paths start to differ wildly from actual filesystem paths: `directoryName/./guid`